source/SkiaSharp.Views/SkiaSharp.Views.WPF/SKGLElement.cs

```diff
@@ -0,0 +1,274 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Media;
+using System.Windows.Media.Imaging;
+using OpenTK.Graphics;
+using System.Windows.Media.Media3D;
+using SkiaSharp.Views.Desktop;
+using OpenTK.Wpf;
+using SkiaSharp;
+using OpenTK.Graphics.OpenGL;
+using OpenTK.Platform.Windows;
+using OpenTK;
+using System.Windows.Interop;
+using OpenTK.Platform;
+#if NETCOREAPP || NET
+using OpenTK.Mathematics;
+#endif
+
+namespace SkiaSharp.Views.WPF
+{
+	[DefaultEvent("PaintSurface")]
+	[DefaultProperty("Name")]
+	public class SKGLElement : GLWpfControl, IDisposable
+	{
+		private const SKColorType colorType = SKColorType.Rgba8888;
+		private const GRSurfaceOrigin surfaceOrigin = GRSurfaceOrigin.BottomLeft;
+
+		private bool designMode;
+
+		private GRContext grContext;
+		private GRGlFramebufferInfo glInfo;
+		private GRBackendRenderTarget renderTarget;
+		private SKSurface surface;
+		private SKCanvas canvas;
+
+		private SKSizeI lastSize;
+		private SKGLElementWindowListener listener;
+
+		public SKGLElement()
+			: base()
+		{
+			Initialize();
+		}
+
+		private void Initialize()
+		{
+            designMode = DesignerProperties.GetIsInDesignMode(this);
+            var settings = new GLWpfControlSettings() { MajorVersion = 2, MinorVersion = 1, RenderContinuously = false };
+			
+            this.Render += OnPaint;
+
+            this.Loaded += (s, e) =>
+            {
+				SKGLElementWindowListener listener = new SKGLElementWindowListener(this);
+				this.listener = listener;
+            };
+
+#if NETCOREAPP
+			this.RegisterToEventsDirectly = false;
+#endif
+
+			Start(settings);
+		}
+
+
+		private class SKGLElementWindowListener: IDisposable
+        {
+            private WeakReference<SKGLElement> toDestroy;
+			private WeakReference<Window> theWindow;
+			public SKGLElementWindowListener(SKGLElement toDestroy)
+            {
+                this.toDestroy = new WeakReference<SKGLElement>(toDestroy);
+				var window = System.Windows.Window.GetWindow(toDestroy);
+				if (window != null)
+				{
+					theWindow = new WeakReference<Window>(window);
+					window.Closing += Window_Closing;
```

Comment:
I see closing that does a hard dispose. Maybe the logic should be in Unloaded and then release the things? At least we can stop the rendering if it is no longer in the UI.

Not sure if closing a window will also trigger the unload.

Reply:
Yeah you've got a point... Normally it's the kind of thing thatd make sense for unload.ight have been working around a problem, been a while now. I can try shifting the cleanup to unload and see if I hit an issue.